### PR TITLE
Issue/7081 Post Editor: Import media from other applications

### PR DESCRIPTION
**Pull request by ScoutHarris**
This partially addresses [7081](https://github.com/wordpress-mobile/WordPress-iOS/issues/7081).

For iOS11+, this feature adds a new 'Other Apps' option to the Post editor > add media functionality by utilizing `UIDocumentPickerViewController` to enable users to add photos and video from other applications.

editor_simulator

**To test:**
- Prerequisite: any device must have external media to select from. Ex: iCloud enabled, Google Drive installed.
- Run the app on iOS11+ devices. Edit a post. Tap the add media button and verify 'Other Apps' appears the menu.
- Verify you can add photos and videos from a selected location.
- Run the app on <iOS11 devices. Verify 'Other Apps' does not appear in the menu.

Needs review: @frosty

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 1]
```diff
@@ -556,6 +568,43 @@ - (void)configureMediaUpload
...
+        }
+    }];
+}
+
+- (void)completeCreateMedia:(Media *)media error:(NSError *)error mediaID:(NSString *)mediaID
+{
+    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
+        if (error || !media || !media.absoluteLocalURL) {
+            [self.editorView removeImage:mediaID];
+            [self.editorView removeImage:mediaID];
```
Comment (catehstn) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: duplication of prev line

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 2]
```diff
@@ -75,7 +75,8 @@ @interface WPPostViewController () <
 UITextViewDelegate,
 UIViewControllerRestoration,
 EditImageDetailsViewControllerDelegate,
-MediaProgressCoordinatorDelegate
+MediaProgressCoordinatorDelegate,
```
Comment (catehstn) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: This file is for the Hybrid editor, you'll want to make this change in Aztec instead.

Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: @catehstn - I'm doing a separate branch for the Aztec toolbar.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 3]
```diff
@@ -2041,6 +2079,15 @@ - (void)editorDidPressMedia:(WPEditorViewController *)editorController
                                       [weakSelf showMediaPickerForWordPressLibraryAnimated:YES];
                                   }];
 
+    if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 11.0 &&
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: I believe that in Xcode 9 you can now write this:

```if (@available(iOS 11, *)) {}```

File: WordPress/Classes/ViewRelated/Reader/WPImageViewController.m [hunk 4]
```diff
@@ -314,7 +314,7 @@ + (BOOL)isUrlSupported:(NSURL *)url
 
     NSArray *types = @[@".png", @".jpg", @".gif", @".jpeg"];
     for (NSString *type in types) {
-        if (NSNotFound != [absoluteURL rangeOfString:type].location) {
+        if (NSNotFound != [[absoluteURL lowercaseString] rangeOfString:type].location) {
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Reader/WPImageViewController.m]: Is this change related to this PR?

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 5]
```diff
@@ -873,6 +921,23 @@ - (void)showMediaPickerWithDataSource:(id<WPMediaCollectionDataSource>)dataSourc
     [self presentViewController:picker animated:animated completion:nil];
 }
 
+- (void)showDocumentPickerAnimated:(BOOL)animated
+{
+    NSArray *docTypes = @[[NSString stringWithFormat:@"%@", kUTTypeImage],
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Can't you simple cast the constant to NSString?

File: WordPress/Classes/ViewRelated/Reader/WPImageViewController.m (hunk 4 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Reader/WPImageViewController.m]: Yes, it is. An image I tried upload had a capitalized extension so it failed this test.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 3 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: @SergioEstevao isn't that Swift? I'm not seeing anything akin to that for ObjC.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 6]
```diff
@@ -873,6 +921,23 @@ - (void)showMediaPickerWithDataSource:(id<WPMediaCollectionDataSource>)dataSourc
...
+                          [NSString stringWithFormat:@"%@", kUTTypeMovie]];
+    UIDocumentPickerViewController *docPicker = [[UIDocumentPickerViewController alloc]
+                                                 initWithDocumentTypes:docTypes inMode:UIDocumentPickerModeImport];
+    docPicker.delegate = self;
+    
+    // The app's appearance settings override the doc picker color scheme.
+    // Setting the nav colors here so the doc picker has the correct appearance.
+    // The app colors are restored later with resetNavigationColors().
+    [[UINavigationBar appearance] setTintColor:[WPStyleGuide mediumBlue]];
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: I think now we have two places where we set these colours (here and in the media library) we could probably extract these into a `WPStyleGuide` extension / method that we call from both places. Then, if ever we need to change this colour, we just have to do it in one place.

Check out our other WPStyleGuide extensions for an example :)

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 3 above)
Comment (SergioEstevao) [reply WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: give it a try :)

File: WordPress/Classes/ViewRelated/Reader/WPImageViewController.m (hunk 4 above)
Comment (SergioEstevao) [reply WordPress/Classes/ViewRelated/Reader/WPImageViewController.m]: Good catch!

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 3 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Well, I did. :) But I'll try again.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 7]
```diff
@@ -556,6 +568,42 @@ - (void)configureMediaUpload
     self.mediaProgressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleBar];
 }
 
+- (void)createMediaThumbnailCallback:(NSURL *)thumbnailURL mediaID:(NSString *)mediaID isImage:(BOOL)isImage
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: I would probably remove `callback` from the name of this method. Maybe something like `handleThumbnailURL:mediaID:isImage:`?

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 8]
```diff
@@ -556,6 +568,42 @@ - (void)configureMediaUpload
...
+
+- (void)completeCreateMedia:(Media *)media error:(NSError *)error mediaID:(NSString *)mediaID
+{
+    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
+        if (error || !media || !media.absoluteLocalURL) {
+            [self.editorView removeImage:mediaID];
+            [self.editorView removeVideo:mediaID];
+            [self stopTrackingProgressOfMediaWithId:mediaID];
+            [WPError showAlertWithTitle:NSLocalizedString(@"Failed to export media",
+                                                          @"The title for an alert that says to the user the media (image or video) he selected couldn't be used on the post.")
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Could we switch 'he' to 'they'?

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 9]
```diff
@@ -556,6 +568,42 @@ - (void)configureMediaUpload
...
+    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
+        if (isImage) {
+            [self.editorView insertLocalImage:thumbnailURL.path uniqueId:mediaID];
+        } else {
+            [self.editorView insertInProgressVideoWithID:mediaID usingPosterImage:thumbnailURL.path];
+        }
+    }];
+}
+
+- (void)completeCreateMedia:(Media *)media error:(NSError *)error mediaID:(NSString *)mediaID
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: I'd perhaps rename this one too... `handleCreatedMedia:`? `handleNewMedia:`? Something like that?

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 10]
```diff
@@ -1850,48 +1915,21 @@ - (void)addDeviceMediaAsset:(PHAsset *)asset
 {
     MediaService *mediaService = [[MediaService alloc] initWithManagedObjectContext:[[ContextManager sharedInstance] mainContext]];
     __weak __typeof__(self) weakSelf = self;
+    bool isImage = (asset.mediaType == PHAssetMediaTypeImage);
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Obj-c should use `BOOL`, not `bool`

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 11]
```diff
@@ -1850,48 +1915,21 @@ - (void)addDeviceMediaAsset:(PHAsset *)asset
...
     NSString *mediaUniqueID = [self uniqueIdForMedia];
     [mediaService createMediaWithPHAsset:asset
                          forPostObjectID:self.post.objectID
                        thumbnailCallback:^(NSURL *thumbnailURL) {
                            __typeof__(self) strongSelf = weakSelf;
-                           if (!strongSelf) {
-                               return;
+                           if (strongSelf) {
+                               [strongSelf createMediaThumbnailCallback:thumbnailURL mediaID:mediaUniqueID isImage:isImage];
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: I _think_ we can just not worry about `strongSelf` here and instead just call this method on `weakSelf`. If self is `nil`, then it'll just fail silently.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 12]
```diff
@@ -1850,48 +1915,21 @@ - (void)addDeviceMediaAsset:(PHAsset *)asset
...
                            }
-                           [[NSOperationQueue mainQueue] addOperationWithBlock:^{
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Nice going pulling out these sections so they're not duplicated for the media and document pickers 👍

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 13]
```diff
@@ -1850,48 +1915,21 @@ - (void)addDeviceMediaAsset:(PHAsset *)asset
...
-                                                 withProperties:[WPAppAnalytics propertiesFor:media]
-                                                       withPost:self.post];
-                                      } else if (media.mediaType == MediaTypeVideo) {
-                                          [WPAppAnalytics track:WPAnalyticsStatEditorAddedVideoViaLocalLibrary
-                                                 withProperties:[WPAppAnalytics propertiesFor:media]
-                                                       withPost:self.post];
-                                      }
-                                      [strongSelf uploadMedia:media trackingId:mediaUniqueID];                                      
-                                  }];
+                                  __typeof__(self) strongSelf = weakSelf;
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: See my previous comment about `strongSelf`.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 3 above)
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: We can now do this the nicer Swiftier way: 

`if (@available(iOS 11.0, *))`

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 14]
```diff
@@ -2312,6 +2359,43 @@ - (BOOL)mediaPickerController:(WPMediaPickerViewController *)picker shouldSelect
...
+    MediaService *mediaService = [[MediaService alloc] initWithManagedObjectContext:[[ContextManager sharedInstance] mainContext]];
+    
+    NSString *mediaUniqueID = [url lastPathComponent];
+    BOOL isImage = [WPImageViewController isUrlSupported:url];
+    __weak __typeof__(self) weakSelf = self;
+    [mediaService createMediaWithURL:url
+                     forPostObjectID:self.post.objectID
+                   thumbnailCallback:^(NSURL *thumbnailURL) {
+                       __typeof__(self) strongSelf = weakSelf;
+                       if (strongSelf) {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: As mentioned above, I don't think we need to worry about strong self here. (someone else please tell me if I'm wrong!)

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 15]
```diff
@@ -2312,6 +2359,43 @@ - (BOOL)mediaPickerController:(WPMediaPickerViewController *)picker shouldSelect
...
+                           [strongSelf createMediaThumbnailCallback:thumbnailURL mediaID:mediaUniqueID isImage:isImage];
+                       }
+                   }
+                          completion:^(Media *media, NSError *error){
+                              __typeof__(self) strongSelf = weakSelf;
+                              if (strongSelf) {
+                                  [strongSelf completeCreateMedia:media error:error mediaID:mediaUniqueID];
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: See strong self comments above.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 3 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Nice hint @SergioEstevao.  Just to confirm... it's working for me as well.

**Comment by ScoutHarris** [comment]
@frosty @SergioEstevao @diegoreymendez all requested changes have been made. Thanks for the feedback!
